modules/programs/command-not-found/command-not-found.nix

```diff
@@ -0,0 +1,85 @@
+# Adapted from Nixpkgs.
+
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.programs.command-not-found;
+  commandNotFound = pkgs.substituteAll {
+    name = "command-not-found";
+    dir = "bin";
+    src = ./command-not-found.pl;
+    isExecutable = true;
+    inherit (pkgs) perl;
+    inherit (cfg) dbPath;
+    perlFlags = concatStrings (map (path: "-I ${path}/lib/perl5/site_perl ")
+      [ pkgs.perlPackages.DBI pkgs.perlPackages.DBDSQLite pkgs.perlPackages.StringShellQuote ]);
+  };
+
+in
+
+{
+  options.programs.command-not-found = {
+
+    enable = mkEnableOption "command-not-found hook for interactive shell";
+
+    dbPath = mkOption {
+      default = "/nix/var/nix/profiles/per-user/root/channels/nixos/programs.sqlite" ;
+      description = ''
+        Absolute path to programs.sqlite.
+
+        By default this file will be provided by your channel
+        (nixexprs.tar.xz).
+      '';
+      type = types.path;
+    };
+  };
+
+  config = mkIf cfg.enable {
+    programs.bash.initExtra =
+      ''
+        # This function is called whenever a command is not found.
+        command_not_found_handle() {
+          local p=${commandNotFound}/bin/command-not-found
+          if [ -x $p -a -f ${cfg.dbPath} ]; then
+            # Run the helper program.
+            $p "$@"
+            # Retry the command if we just installed it.
+            if [ $? = 126 ]; then
+              "$@"
+            else
+              return 127
+            fi
+          else
+            echo "$1: command not found" >&2
+            return 127
+          fi
+        }
+      '';
+
+    programs.zsh.initExtra =
+      ''
+        # This function is called whenever a command is not found.
+        command_not_found_handler() {
+          local p=${commandNotFound}/bin/command-not-found
+          if [ -x $p -a -f ${cfg.dbPath} ]; then
+            # Run the helper program.
+            $p "$@"
+
+            # Retry the command if we just installed it.
+            if [ $? = 126 ]; then
+              "$@"
+            fi
```

Comment:
Hmm, do you know if there is some specific reason why it doesn't return 127 for Zsh? If it did then it seems to me that the code could be
```
programs.bash.initExtra = shInit "command_not_found_handle";
programs.zsh.initExtra = shInit "command_not_found_handler";
```
with `shInit` returning these strings…

Reply:
Looking at the history of the original module it looks like the part of the code was just forgotten to be added in both shells, so I think we can merge them. However, this branch is used when `NIX_AUTO_INSTALL` is set and perl script will automatically install missing packages via `nix-env -iA`. I'm thinking if we should remove this branch completely from home-manager module implementation since it is opposite of what we are trying to do here -- eliminate usage of `nix-env` command and make the home configuration declarative. @rycee what do you think?

Reply:
Ya, I agree with that. I'm not very fond of the the idea of `NIX_AUTO_INSTALL` so I don't mind at all seeing it gone :-)